### PR TITLE
Add support for git_pillar 'mountpoint' parameter.

### PR DESCRIPTION
**Pull request by mrichar1**
### What does this PR do?

This PR attempts to provide a solution to the request for `mountpoint` support in git_pillar.
### What issues does this PR fix or reference?

This behaviour is requested in #34210 
### New Behavior

When mountpoint is defined, the code now creates a symlink of the original repo cache into a new directory, linked at the mountpoint. This is then used as the path form which the pillar config is read.

For example:

```
ext_pillar:
  - git:
    - master https://git.example.com/top.git:
    - master https://git.example.com/devolved.git
      - mountpoint: /devolved
```

This will create:

```
/var/cache/salt/git_pillar/{{hash_of_top}} # Added to pillar list
/var/cache/salt/git_pillar/{{hash_of_devolved}} # Now ignored
/var/cache/salt/git_pillar/links/{{hash_of_devolved}} # Added to pillar list
/var/cache/salt/git_pillar/links/{{hash_of_devolved}}/devolved # Symlink to 'real' repo
```
### Tests written?

Not yet - let me know if this approach is a viable route, and if any of the code needs changing. It might be advisable to use a different option than `mountpoint` since this behaviour is subtly different to that of `mountpoint` in `gitfs_remotes` and might cause confusion.

### CODE REVIEW

**Comment by cachedout** [comment]
@terminalmage Any chance you could jump in here?

**Comment by terminalmage** [comment]
I'm not sure this will work, if I remember correctly each repo is processed separately. I don't have time to test this right now, I will try to test by the end of the day on Thursday.

**Comment by mrichar1** [comment]
I've been running this patch on a test salt master and it all seems to work fine (though I am aware that I might not be hitting all use cases and code paths).

The patch tries to mimic the original behaviour as much as possible, just changing the 'base path' from which the cloned repos are considered.

Say for example there are 4 repos, 2 with mountpoint `/` (`a` and `b`) and 2 with mountpoint `/devolved` (`c` and `d`).

This patch will create (using names instead of hashes for simplicity here):

```
/repos/a
/repos/b
/repos/c
/repos/d
/links/c/devolved
/links/d/devolved
```

It will then set `self.pillar_dirs` to contain:

```
/repos/a
/repos/b
/links/c
/links/d
```

The merge code will then end up with a final tree containing:

```
/ # in a and b
/devolved # in c and d
```

For an extreme example, if `a` and `b` contained `/devolved/foo.sls` and `c` and `d` contained `/foo.sls` they would all be merged together as per the merge ordering code. The controller of the master config can decide what order to include the gitfs remotes, and thus change the merging.

I'll see about doing some more tests on this (in particular the `root` option, which I've not looked at) but hopefully the philosophy behind it is good, even if the patch needs improvements!

**Comment by cachedout** [comment]
Bump @terminalmage on a follow-up here please.

**Comment by cachedout** [comment]
@terminalmage Could you let me know on this one, please? Thanks.

File: salt/utils/gitfs.py [hunk 1]
```diff
@@ -2920,6 +2920,9 @@ def checkout(self):
         Checkout the targeted branches/tags from the git_pillar remotes
         '''
         self.pillar_dirs = {}
+        linkdir = os.path.join(self.cache_root, 'links')
+        if not os.path.isdir(linkdir):
+            os.makedirs(linkdir)
```
Comment (isbm) [inline salt/utils/gitfs.py]: And what happens if file?

File: salt/utils/gitfs.py [hunk 2]
```diff
@@ -2929,7 +2932,17 @@ def checkout(self):
                 else:
                     base_branch = self.opts['{0}_base'.format(self.role)]
                     env = 'base' if repo.branch == base_branch else repo.branch
-                self.pillar_dirs[cachedir] = env
+                if repo.mountpoint:
+                    lcachedir = os.path.join(linkdir, repo.hash)
+                    if not os.path.isdir(lcachedir):
```
Comment (isbm) [inline salt/utils/gitfs.py]: I'd rather check if `os.path.exists`.

**Review by isbm — CHANGES_REQUESTED** [review]

**Comment by terminalmage** [comment]
@mrichar1 it turns out that some changes that were made in 05782aa do allow a git_pillar repo to see other repos the same ``git`` section in the ``ext_pillar`` configuration. However, this comes with a couple caveats:

1. This is limited to a single git_pillar environment.
2. All git_pillar repos from each environment are still processed separately, they just see the files from the other repos in the same environment.

So, this code may technically work, but it would be inefficient since it would still try to compile pillar data for the mounted repo. Also, it might be desirable to disable this "include" functionality added in 05782aa.

What I'm going to do is take the work that you started and build upon it to address the above concerns, and any other issues I identify along the way. I will then open my own PR which will contain the combined work from @mrichar1 and myself.

**Comment by mrichar1** [comment]
Sounds great - thanks!

**Comment by terminalmage** [comment]
@mrichar1 No problem, thanks for getting this started. I'm sorry it took so long to get to this, but I'm pretty much the only one that fully groks the salt.utils.gitfs stuff since I was the one that wrote it, so it wasn't something that another engineer could properly review. And things have just been super busy these past several months. I'm looking forward to getting this in though, I hope to have it completed by the end of the week or early next week. I've already written a ton of new documentation to explain how the mountpoint feature and the ability to include files from other repos in the same environment, and I've begun work on the changes and optimizations as welll.

**Comment by terminalmage** [comment]
@mrichar1 Hey, just letting you know that this is still in progress. My whole family got sick and I've also been working on high-priority features/bugfixes, but I am mostly done with the modifications that needed to be made. This is still on target to be completed for the next feature release.

**Comment by terminalmage** [comment]
Back to working on this after having to put it on the back-burner again.

**Comment by terminalmage** [comment]
https://github.com/saltstack/salt/pull/39629 supersedes this PR and completes this feature. Closing this PR in favor of https://github.com/saltstack/salt/pull/39629.

Thanks again, @mrichar1, for doing the initial investigation and work on this. I'm sorry it took so long to get this done.

**Pull request closed without merging**